### PR TITLE
Display span information on Z page

### PR DESCRIPTION
**Pull request by n1dtec**

### CODE REVIEW

**Comment by MikeGoldsmith** [comment]
Hey @n1dtec - Is this ready for review, or just open to help visualise the PRs changes?

**Comment by n1dtec** [comment]
@MikeGoldsmith Hi, This is not ready for review yet.

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs [hunk 1]
```diff
@@ -121,9 +121,13 @@ private void WorkerThread()
                         using (var writer = new StreamWriter(output))
                         {
                             writer.WriteLine("Span Count : " + this.spanProcessor.GetSpanCount());
+                            writer.WriteLine("Error Count : " + this.spanProcessor.GetErrorCount());
+                            writer.WriteLine("Span Name\tStatus\t \tStart Timestamp\t \t \tEnd Timestamp\t \t \tLatency");
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs]: please use `<tr><td>` syntax to build a table. So we will be able to apply CSS styles to this page.

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs [hunk 2]
```diff
@@ -121,9 +121,13 @@ private void WorkerThread()
                         using (var writer = new StreamWriter(output))
                         {
                             writer.WriteLine("Span Count : " + this.spanProcessor.GetSpanCount());
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs]: Global count is not an interesting metric. Please start moving it under the specific span name

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs [hunk 3]
```diff
@@ -121,9 +121,13 @@ private void WorkerThread()
...
+                            writer.WriteLine("Error Count : " + this.spanProcessor.GetErrorCount());
+                            writer.WriteLine("Span Name\tStatus\t \tStart Timestamp\t \t \tEnd Timestamp\t \t \tLatency");
                             foreach (var spanData in this.spanProcessor.GetSpanDataBatch())
                             {
-                                writer.WriteLine("Span Name : " + spanData.Name + "\t" + "Status : " + spanData.Status.CanonicalCode);
+                                long latency = spanData.EndTimestamp.ToUnixTimeMilliseconds() > spanData.StartTimestamp.ToUnixTimeMilliseconds() ? spanData.EndTimestamp.ToUnixTimeMilliseconds() - spanData.StartTimestamp.ToUnixTimeMilliseconds() : 0;
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs]: you don't have to hide the negative values. Those will indicate an issue with the data collection that will be easier to discover with zPages

File: src/OpenTelemetry/Trace/Export/SimpleSpanProcessor.cs [hunk 4]
```diff
@@ -28,6 +30,7 @@ public class SimpleSpanProcessor : SpanProcessor
         private readonly SpanExporter exporter;
         private bool disposed = false;
         private long spanCount = 0;
+        private List<SpanData> batch = new List<SpanData>();
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry/Trace/Export/SimpleSpanProcessor.cs]: instead of ever-growing list that may cause Out of Memory exception - please track the metrics for the last minute and last hour

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace OpenTelemetry.Exporter.ZPages.Implementation
+{
+    internal class ZPagesSpanInformation
+    {
+        public ZPagesSpanInformation()
+        {
+        }
+
+        public ZPagesSpanInformation(SpanData spanData)
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs]: why `ZPagesSpanInformation` has a `spanData` as an argument?

Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs]: Perhaps it must be a name of it

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs [hunk 6]
```diff
@@ -29,6 +31,8 @@ namespace OpenTelemetry.Exporter.ZPages
     public class ZPagesExporter : SpanExporter
     {
         internal readonly ZPagesExporterOptions Options;
+        private Dictionary<string, ZPagesSpanInformation> spanList = new Dictionary<string, ZPagesSpanInformation>();
+        private IEnumerable<SpanData> batch = new List<SpanData>();
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs]: this list is always growing. Will lead to memory leak

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs [hunk 7]
```diff
@@ -42,6 +46,30 @@ public ZPagesExporter(ZPagesExporterOptions options)
...
+                    int count = this.spanList.Count;
+                }
+                else
+                {
+                    ZPagesSpanInformation spanInformation;
+                    this.spanList.TryGetValue(spanData.Name, out spanInformation);
+                    if (spanInformation != null)
+                    {
+                        spanInformation.SpanDataList.Add(spanData);
+                        spanInformation.CountHour++;
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs]: hm, how would it work? When will you reset the counter?

**Comment by cijothomas** [comment]
@n1dtec Are you planning to work on this PR? As part of overall direction of moving out all exporters other than Jaeger,Zipkin, & Prometheus, I'd like to remove ZPagesExporter as well. 
I think we can move it to contrib repo, and whenever you are ready, you can submit this PR in that repo.
https://github.com/open-telemetry/opentelemetry-dotnet-contrib

**Comment by SergeyKanzhelev** [comment]
I spoke to @n1dtec and asked to proceed with it here to ensure it works. Also I want to bring zPages to specs and suggest to make it a part of a default offering.

We can discuss it on SIG meeting,

@n1dtec any updates on this PR?

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+        /// <summary>
+        /// Gets or sets the total error count of the span.
+        /// </summary>
+        public long ErrorTotal { get; set; }
+
+        /// <summary>
+        /// Gets or sets the total average latency of the span.
+        /// </summary>
+        public long AvgLatencyTotal { get; set; }
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs:75]: Why latency is counted as `long`? What the unit of measurement? Perhaps - use `Timespan` type instead of `long`?

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        private static long startTime;
+
+        /// <summary>
+        /// Initializes static members of the <see cref="ZPagesSpans"/> class.
+        /// </summary>
+        static ZPagesSpans()
+        {
+            ZQueue = new LinkedList<Dictionary<string, ZPagesSpanInformation>>();
+            ProcessingSpanList = new Dictionary<string, long>();
+            CurrentMinuteSpanList = new Dictionary<string, ZPagesSpanInformation>();
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs:38]: this either needs to be a concurrent dictionary or have lock on access

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs [hunk 10]
```diff
@@ -13,10 +13,16 @@
...
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Timers;
+using OpenTelemetry.Exporter.ZPages.Implementation;
 using OpenTelemetry.Trace.Export;
+using Timer = System.Timers.Timer;
 
 namespace OpenTelemetry.Exporter.ZPages
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs:27]: What's the role of exporter? It is not shown on specification: https://docs.google.com/document/d/1vgdbaL7Qb11DFBiF3PzZS9KUJ24VqJuBkKMq3oB-yJQ/edit#

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs [hunk 11]
```diff
@@ -34,11 +42,22 @@ public class ZPagesExporter : SpanExporter
         public ZPagesExporter(ZPagesExporterOptions options)
         {
             this.Options = options;
+
+            // Create a timer with one minute interval
+            this.minuteTimer = new Timer(60000);
+            this.minuteTimer.Elapsed += new ElapsedEventHandler(ZPagesSpans.TriggerMinuteCalculations);
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs:48]: In a specification the only action here is Purge of stale data. Should we call it Purge than?

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        /// <summary>
+        /// Triggers Calculations every minute.
+        /// </summary>
+        /// <param name="source">Source.</param>
+        /// <param name="e">Event Arguments.</param>
+        public static void TriggerMinuteCalculations(object source, ElapsedEventArgs e)
+        {
+            // Enqueue the last minute's span information list
+            ZQueue.AddFirst(CurrentMinuteSpanList);
+            CurrentMinuteSpanList.Clear();
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs:101]: I don't understand. You are cleaning the dictionary that have just been queued for aggregation... So you wouldn't be aggregating anything at all

Comment (n1dtec) [reply src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs:101]: When a new minute starts, the dictionary gets cleared. So it will start storing information only about new spans. If the queued information needs to be accessed, it is at the front of the queue.

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs (hunk 8 above)
Comment (n1dtec) [reply src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpanInformation.cs:75]: The unit for latency is milliseconds, hence it was counted as long. The same gets displayed on the HTML. Do we need to use TimeSpan instead?

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs [hunk 13]
```diff
@@ -108,17 +110,69 @@ private void WorkerThread()
...
+                                             "<script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js\"></script>" +
+                                             "<script src=\"https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js\"></script></head>");
+                            writer.WriteLine("<body><div class=\"col-sm-1\"></div><div class=\"container col-sm-10\"><div class=\"jumbotron table-responsive\"><h1>RPC Stats</h2>" +
+                                             "<table class=\"table table-bordered table-hover table-striped\">" +
+                                             "<thead style=\"color: white;background-color: Teal;\"><tr><th>Span Name</th><th>Total Count</th><th>Count in last minute</th><th>Count in last hour</th><th>Average Latency (ms)</th>" +
+                                             "<th>Average Latency in last minute (ms)</th><th>Average Latency in last hour (ms)</th><th>Total Errors</th><th>Errors in last minute</th><th>Errors in last minute</th><th>Last Updated</th></tr></thead>" +
+                                             "<tbody style=\"background-color: white;\">");
```
[the remark this answers is not in the document]
Comment (n1dtec) [reply src/OpenTelemetry.Exporter.ZPages/ZPagesExporterStatsHttpServer.cs]: I have removed the explicit styling for now. Adding a css file in the code did not work since the HTML looks for a css which is hosted on the server, instead of finding it in the project.
image

File: src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs (hunk 10 above)
Comment (n1dtec) [reply src/OpenTelemetry.Exporter.ZPages/ZPagesExporter.cs:27]: I have added it now in the In-Process zPages implementation section

File: src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+        /// <summary>
+        /// Gets or sets the total latency of spans name-wise.
+        /// </summary>
+        public static Dictionary<string, long> TotalSpanLatency { get; set; }
+
+        /// <summary>
+        /// Gets or sets the retention time (in milliseconds) for the metrics.
+        /// </summary>
+        public static long RetentionTime { get; set; }
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry.Exporter.ZPages/Implementation/ZPagesSpans.cs:90]: Maybe this could be a `TimeSpan` instead of long? That will allow you set it using TimeSpan.FromXXX utility functions (eg `TimeSpan.FromSeconds(60)`) and time comparisons do not need to be converted using `ToUnixTimeMilliseconds`.
